### ISSUE TITLE
Architecture: AST Enrichment Pipeline - Models as Thought Accumulators

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#106]
## Vision

Transform Klareco's architecture from a monolithic model to a **pipeline of specialized micro-models**, where the AST acts as a structured "thought" that accumulates semantic meaning as it passes through each stage.

```
Text → Parser → AST₀ → Semantic → AST₁ → Grammatical → AST₂ → Discourse → AST₃ → Reasoning → ASTₙ
              │        Model      │        Model        │        Model       │        Model
              │                   │                     │                    │
         Structure           + Content             + Feature            + Coref
         (grammar)            meaning               effects              links
```

## Core Insight

Each model:
1. **Reads** the current AST (including annotations from previous models)
2. **Adds** its specialized semantic annotations
3. **Passes** the enriched AST to the next model

The AST becomes a "thought" that grows more complete at each stage.

### 1. Leverages Klareco's Core Thesis
- Parser already extracts structure deterministically
- Each learned model focuses ONLY on semantics it can't get from grammar
- No wasted parameters on structure detection

### 2. Extreme Modularity
| Model | Parameters | Learns |
|-------|------------|--------|
| Parser | 0 | (Deterministic) |
| Semantic | ~323K | Root/affix meanings |
| Grammatical | ~32K | Negation, tense, mood effects |
| Discourse | ~100K | Coreference, connectives |
| **Pre-reasoning total** | **~455K** | Full sentence understanding |

### 3. Full Interpretability
Inspect AST at any stage:
```python
# After semantic model
ast.content_embeddings['hundo'] = [0.2, -0.5, ...]

# After grammatical model
ast.negation_effect = tensor([...])
ast.temporal_reference = 'completed_past'

# After discourse model
ast.coreference_links['li'] = 'zamenhof_entity_42'
```

### 4. Independent Training
Train each model separately on its specific task:
- Semantic model: root/affix similarity pairs
- Grammatical model: minimal pairs (same content, different grammar)
- Discourse model: coreference chains, connective relations

### 5. Composable Pipelines
Different tasks use different subsets:
- **Retrieval**: Parser → Semantic → Index
- **Q&A**: Parser → Semantic → Grammatical → Discourse → Reasoning
- **Analysis**: Parser → Semantic → Grammatical (inspect)

## AST Schema with Semantic Slots

```python
@dataclass
class EnrichedAST:
    # Stage 0: Parser (deterministic)
    tipo: str
    subjekto: Optional[WordGroup]
    verbo: Optional[Word]
    objekto: Optional[WordGroup]
    aliaj: List[Word]
    negita: bool
    tempo: str
    modo: str
    fraztipo: str
    
    # Stage 1: Semantic Model (learned)
    content_embeddings: Dict[str, Tensor] = None
    sentence_embedding: Optional[Tensor] = None
    
    # Stage 2: Grammatical Semantics Model (learned)
    grammatical_embedding: Optional[Tensor] = None
    negation_interpretation: Optional[str] = None
    temporal_reference: Optional[str] = None
    modality: Optional[str] = None
    
    # Stage 3: Discourse Model (learned)
    coreference_links: Dict[str, str] = None
    discourse_relation: Optional[str] = None
    
    # Stage 4: Reasoning Model (learned)
    inferences: List[EnrichedAST] = None
    confidence: float = None
```

### Semantic Model (~323K params)
```python
class SemanticModel(nn.Module):
    def __init__(self, dim=64):
        self.root_embed = nn.Embedding(num_roots, dim)
        self.affix_embed = nn.Embedding(num_affixes, dim)
    
    def forward(self, ast: EnrichedAST) -> EnrichedAST:
        for word in ast.content_words():
            ast.content_embeddings[word.id] = self.embed_word(word)
        ast.sentence_embedding = self.pool(ast.content_embeddings)
        return ast
```

### Grammatical Semantics Model (~32K params)
```python
class GrammaticalModel(nn.Module):
    def __init__(self, dim=64):
        # Dedicated transform for each grammatical feature
        self.negation = nn.Linear(dim, dim)
        self.past = nn.Linear(dim, dim)
        self.future = nn.Linear(dim, dim)
        self.conditional = nn.Linear(dim, dim)
        self.question = nn.Linear(dim, dim)
        # ... ~8 features × 4K params each
    
    def forward(self, ast: EnrichedAST) -> EnrichedAST:
        emb = ast.sentence_embedding
        
        # AST tells us WHAT to apply, we learn WHAT IT MEANS
        if ast.negita:
            emb = emb + self.negation(emb)
        if ast.tempo == 'pasinteco':
            emb = emb + self.past(emb)
        # ...
        
        ast.grammatical_embedding = emb
        return ast
```

### Pipeline Executor
```python
class KlarecoPipeline:
    def __init__(self):
        self.semantic = SemanticModel()
        self.grammatical = GrammaticalModel()
        self.discourse = DiscourseModel()
    
    def __call__(self, text: str, context: List[EnrichedAST] = None):
        ast = parse(text)                    # Stage 0: Structure
        ast = self.semantic(ast)             # Stage 1: Content meaning
        ast = self.grammatical(ast)          # Stage 2: Feature effects
        if context:
            ast = self.discourse(ast, context)  # Stage 3: Cross-sentence
        return ast
```

### Stage 1: Semantic Model
- **Data**: Root similarity pairs, definition overlap
- **Loss**: Contrastive on content embeddings
- **Freeze after**: Yes

### Stage 2: Grammatical Model  
- **Data**: Minimal pairs (same content, different grammar)
- **Loss**: Similarity targets for grammatical contrasts
- **Uses**: Frozen semantic model for base embeddings
- **Freeze after**: Yes

### Stage 3: Discourse Model
- **Data**: Coreference chains, discourse relations
- **Loss**: Ranking loss for coreference, classification for relations
- **Uses**: Frozen semantic + grammatical models
- **Freeze after**: Yes

### Stage 4: Reasoning Model
- **Data**: Q&A pairs, inference chains
- **Uses**: All frozen earlier models
- **Trains**: AST-to-AST transformation

## Benefits Summary

| Aspect | Monolithic | Pipeline |
|--------|------------|----------|
| Parameters | Mixed, hard to budget | Clear per-model budget |
| Training | All at once, complex | Stage by stage, simple |
| Debugging | Black box | Inspect any stage |
| Retraining | Full model | Just affected stage |
| Interpretability | Low | Full AST inspection |
| Composability | One pipeline | Mix and match |

## Analogy: Human Understanding

1. **Perception** (Parser): "I see words, recognize structure"
2. **Lexical Access** (Semantic): "I know what these words mean"  
3. **Grammatical Integration** (Grammatical): "This is negated, past tense"
4. **Discourse Integration** (Discourse): "'He' refers to Zamenhof"
5. **Reasoning** (Reasoning): "Therefore, the answer is..."

The AST accumulates understanding like a thought becoming complete.

## Implementation Plan

1. [ ] Define EnrichedAST schema with all semantic slots
2. [ ] Refactor SemanticModel to write to AST slots
3. [ ] Implement GrammaticalModel with dedicated transforms
4. [ ] Implement Pipeline executor
5. [ ] Update training scripts for staged training
6. [ ] Add AST inspection/visualization tools

## Related Issues

- #101 Grammatical features with semantic content
- #102-105 Individual grammatical feature semantics
- #100 Training Plan v3
- #95 Meta: Semantics layer

## Questions to Resolve

1. Should embeddings be stored in AST as tensors or serialized?
2. How to handle batching across pipeline stages?
3. Should we checkpoint ASTs between stages for debugging?
4. GPU memory management with multiple models?

**Comment by marctjones** [comment]
## Parameter Update: Stage 2 Expanded

After comprehensive audit of grammatical features with semantic content (#101), Stage 2 has been expanded:

### Updated Parameter Budget

| Model | Original | Updated | Change |
|-------|----------|---------|--------|
| Semantic | ~333K | ~333K | - |
| **Grammatical** | **~32K** | **~52K** | **+20K** |
| Discourse | ~100K | ~100K | - |
| **Pre-reasoning** | **~455K** | **~485K** | **+30K** |

### New Grammatical Transforms Added

In addition to original (negation, tense, mood, sentence type, direction):

| Transform | Params | Issue |
|-----------|--------|-------|
| Comparison (pli/plej/ol) | 4K | #108 |
| Aspect (ek-, -ad-) | 4K | #109 |
| Focus particles (nur, eĉ, ankaŭ, ja) | 8K | #110 |
| Evidentiality | 4K | #111 |
| Possessive | 4K | #112 |

### Still Tiny!

~485K params for full sentence understanding (vs 110M for BERT) demonstrates the power of leveraging Esperanto's regular grammar through AST.

See updated `TRAINING_PLAN_V3.md` and wiki page [[AST-Enrichment-Pipeline]].

**Comment by marctjones** [comment]
## Status Update (2025-12-27)

### Stage 1 Semantic Model: ✅ TRAINED (needs refactoring)

**Current implementation:**
- Root embeddings: 11,121 × 64d = 712K params
- Affix embeddings: 43 × 32d = 1.4K params
- FAISS index: 1.94M sentences, 97.7% success rate
- Validation: 5/5 tests pass

**Gap from AST Enrichment architecture:**
The current implementation outputs **flat vectors**, not **annotated ASTs**:

```python
# CURRENT (bypasses AST)
def embed_sentence(text):
    ast = parse(text)
    words = extract_words(ast)  # Flatten!
    return mean([embed_word(w) for w in words])  # Loses structure!

# NEEDED (preserves AST)
def enrich_ast(ast):
    if ast['tipo'] == 'vorto':
        ast['embedding'] = embed_word(ast)  # Annotate in-place
    for child in children(ast):
        enrich_ast(child)
    return ast
```

### Architectural Decision

Current approach works for retrieval (find similar sentences) but **breaks the AST-as-thoughts pipeline**.

**Options:**
1. **Refactor Stage 1** to output enriched ASTs before building Stage 2
2. Keep flat vectors for retrieval, add parallel enriched-AST path
3. Store both in FAISS metadata

**Recommendation:** Option 1 - Refactor now before building more on wrong foundation.

### Implementation Plan Update

- [x] Define EnrichedAST schema with semantic slots (done in #106)
- [ ] **Refactor SemanticModel to write to AST slots** ← NEXT
- [ ] Update FAISS indexing to store enriched ASTs
- [ ] Implement GrammaticalModel (Stage 2)
- [ ] Implement Pipeline executor
- [ ] Add AST inspection/visualization tools

### Related Issues
- #133 (new): Refactor Stage 1 to output enriched ASTs
- #132: Parser bug affects 2.3% of sentences
- #131: Affix architecture evaluation (can be closed - current approach validated)

**Comment by marctjones** [comment]
## Dependencies Update (December 2025)

This issue (#106) is the **parent architecture issue** for the AST enrichment pipeline. It has been broken down into concrete implementation issues:

### Implementation Order

```
#154 EnrichedAST Schema ← Foundation (no dependencies)
         ↓
#155 Pipeline Executor ← Chains models together
         ↓
    ├── #133 Stage 1 Refactor ← Update indexing to use EnrichedAST
    ├── #107 Thought Visualizer ← Debugging/inspection tool
    │        ↓
    └── Stage 2 Issues (#102-105, #108-112) ← Grammatical transforms
              ↓
         Stage 3 Issues (#92, #93, #94) ← Discourse
              ↓
         #156 Answer Composition ← Combine ASTs into answers
              ↓
         Stage 4: Reasoning Core (future)
```

### Dependency Chain

| Issue | Title | Blocked By | Blocks |
|-------|-------|------------|--------|
| #154 | EnrichedAST Schema | None | #155, #133, #107 |
| #155 | Pipeline Executor | #154 | #133, Stage 2 |
| #133 | Stage 1 Refactor | #154, #155 | Corpus reindexing |
| #107 | Thought Visualizer | #155 | (demo/debugging) |
| #156 | Answer Composition | #155, #154 | Stage 4 |

### Status

- [x] Architecture designed (this issue)
- [ ] #154 EnrichedAST implementation
- [ ] #155 Pipeline executor
- [ ] #133 Stage 1 refactor
- [ ] #107 Visualization tools

This meta-issue can be closed when #154, #155, and #133 are complete.
